Optimizely article complete tracking

amoore108 opened the pull request:
Resolves https://github.com/bbc/simorgh/issues/9941

**Overall change:**
Adds a new component after the article body to send an Optimizely 'article_completes' event when scrolled into the viewport

**Code changes:**
- Adds a new component `<OptimizelyArticleCompleteTracking />` below the StoryPage article body. This component makes use of the IntersectionObserver, similar to the `useViewTracker` hook, to send the view event to Optimizely once the component is in view. Similar checks are in place to only fire this event if the user is in an experiment, the component is on screen, its not on AMP and the event hasn't been fired already
- The element uses the `aria-hidden` attribute to remove it from the screen-reader flow

---

- [x] I have assigned myself to this PR and the corresponding issues
- [ ] I have added the `cross-team` label to this PR if it requires visibility across World Service teams
- [x] I have assigned this PR to the Simorgh project
- [x] (BBC contributors only) This PR follows the [repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

**Testing:**

- [ ] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false yarn test:e2e:interactive`)
- [ ] This PR requires manual testing

amoore108 committed:
Initial commit

amoore108 committed:
Update index.jsx

amoore108 committed:
Merge branch 'latest' into optimizely-article-complete-tracking

amoore108 committed:
Update index.jsx

amoore108 committed:
Update index.jsx

amoore108 committed:
Update index.jsx

amoore108 committed:
Update index.test.jsx.snap

amoore108 committed:
Merge branch 'latest' into optimizely-article-complete-tracking

amoore108 committed:
Merge branch 'latest' into optimizely-article-complete-tracking

mfonofm reviewed (commented):
Looks good! Just one query, why are we using `aria-hidden` in the snapshot?

amoore108 commented:
> Looks good! Just one query, why are we using `aria-hidden` in the snapshot?

The `<OptimizelyArticleCompleteTracking />` component returns an empty `<div />` to allow the IntersectionObserver to detect the element on screen. I added in `aria-hidden="true"` just to ensure that the empty div isn't captured by a screen-reader.

andrewscfc commented on src/app/containers/OptimizelyArticleCompleteTracking/index.test.jsx:
```diff
@@ -0,0 +1,211 @@
+
+    await waitFor(() => {
+      expect(optimizely.track).toHaveBeenCalledTimes(0);
+    });
```
I think this test would always pass as it's a negative test. The expectation will instantly pass as it will have initially be called zero times. The only way I can think of to make a test like this work is to add an arbitrary timed wait but even that can be flawed in the logic takes a whole to execute. I'd suggest removing the test on that basis unless you can think of a different approach?

andrewscfc commented on src/app/containers/OptimizelyArticleCompleteTracking/index.test.jsx:
```diff
@@ -0,0 +1,211 @@
+
+    await waitFor(() => {
+      expect(optimizely.track).toHaveBeenCalledTimes(0);
+    });
```
Same with this test, it will instantly pass regardless of the application logic

andrewscfc commented on src/app/containers/OptimizelyArticleCompleteTracking/index.test.jsx:
```diff
@@ -0,0 +1,211 @@
+    });
+
+    await waitFor(() => {
+      expect(global.IntersectionObserver).toHaveBeenCalledTimes(1);
```
In a way you wouldn't expect this to be called on amp as no user js is allowed to run. You might want to go a step further and return null from the the component on amp rather than rendering the div; you could test for this

andrewscfc reviewed (commented):
Looks good, just a few tests that unfortunately may not have that much value, please feel free to disagree!

mfonofm commented:
> > Looks good! Just one query, why are we using `aria-hidden` in the snapshot?
> 
> The `<OptimizelyArticleCompleteTracking />` component returns an empty `<div />` to allow the IntersectionObserver to detect the element on screen. I added in `aria-hidden="true"` just to ensure that the empty div isn't captured by a screen-reader.

Perfect. Sorry, somehow I missed the `aria-hidden` within the component :sweat_smile:

amoore108 committed:
Test updates and return null if on AMP

amoore108 commented:
@andrewscfc Thanks for the comments, I've updated those tests to be more inline with the `useViewTracker` hook tests, checking for fetch calls after a period of time. Also updated the component to return null if on AMP and the test to reflect this.

andrewscfc reviewed (commented):
Thanks for the changes @amoore108 , I'm still not 100% convinced the non-AMP tests would catch a regression where the tracking event is fired when it shouldn't be? Have you tried deliberately firing the tracking event for the two negative tests and seeing whether to tests fail? Tests I'm referring to:
https://github.com/bbc/simorgh/pull/9951/files#diff-16f0ddcf253cc4ca934f810d1008f90350024f9e488435447c73382a4f7a0df3R114
https://github.com/bbc/simorgh/pull/9951/files#diff-16f0ddcf253cc4ca934f810d1008f90350024f9e488435447c73382a4f7a0df3R114

amoore108 committed:
Test fixes

amoore108 commented:
> Thanks for the changes @amoore108 , I'm still not 100% convinced the non-AMP tests would catch a regression where the tracking event is fired when it shouldn't be? Have you tried deliberately firing the tracking event for the two negative tests and seeing whether to tests fail? Tests I'm referring to: https://github.com/bbc/simorgh/pull/9951/files#diff-16f0ddcf253cc4ca934f810d1008f90350024f9e488435447c73382a4f7a0df3R114 https://github.com/bbc/simorgh/pull/9951/files#diff-16f0ddcf253cc4ca934f810d1008f90350024f9e488435447c73382a4f7a0df3R114

Ah, yea you're right. I thought I had tried switching between the `isIntersecting` to check, but mustn't have!

I have updated the tests again and it appears to work correctly now. Switching `isIntersecting` will result in a fail now for that particular test. The problem stemmed from the `optimizely.onReady` promise not being resolved fully before checking the calls had been made.

amoore108 committed:
Merge branch 'latest' into optimizely-article-complete-tracking

DarioR01 reviewed (approved):
LGTM
